src/util/PathUtil.cpp

```diff
@@ -1,63 +1,206 @@
+                        filename.string());
+        XojMsgBox::showErrorToUser(nullptr, msg);
     }
+}
 
-    while (!feof(fpRead)) {
-        size_t bytes = fread(buffer.data(), 1, buffer.size(), fpRead);
-        if (bytes) {
-            fwrite(buffer.data(), 1, bytes, fpWrite);
-        }
+void Util::openFileWithFilebrowser(const fs::path& filename) {
+#ifdef __APPLE__
+    constexpr auto const OPEN_PATTERN = "open \"{1}\"";
+#elif _WIN32
+    constexpr auto const OPEN_PATTERN = "explorer.exe /n,/e,\"{1}\"";
+#else  // linux, unix, ...
+    constexpr auto const OPEN_PATTERN = R"(nautilus "file://{1}" || dolphin "file://{1}" || konqueror "file://{1}" &)";
+#endif
+    string command = FS(FORMAT_STR(OPEN_PATTERN) % Util::getEscapedPath(filename));
+    if (system(command.c_str()) != 0) {
+        string msg = FS(_F("File couldn't be opened. You have to do it manually:\n"
+                           "URL: {1}") %
+                        filename.string());
+        XojMsgBox::showErrorToUser(nullptr, msg);
     }
+}
+
+auto Util::getAutosaveFilepath() -> fs::path {
+    fs::path p(getConfigSubfolder("autosave"));
+    p /= std::to_string(getPid()) + ".xopp";
+    return p;
+}
+
+auto Util::getConfigFolder() -> fs::path {
+    fs::path p(g_get_user_config_dir());
+    p /= g_get_prgname();
+    return p;
+}
+
+auto Util::getConfigSubfolder(const fs::path& subfolder) -> fs::path {
+    fs::path p = getConfigFolder();
+    p /= subfolder;
+
+    return Util::ensureFolderExists(p);
+}
+
+auto Util::getCacheSubfolder(const fs::path& subfolder) -> fs::path {
+    fs::path p(g_get_user_cache_dir());
+    p /= g_get_prgname();
+    p /= subfolder;
+
+    return Util::ensureFolderExists(p);
+}
 
-    fclose(fpRead);
-    fclose(fpWrite);
+auto Util::getDataSubfolder(const fs::path& subfolder) -> fs::path {
+    fs::path p(g_get_user_data_dir());
+    p /= g_get_prgname();
+    p /= subfolder;
+
+    return Util::ensureFolderExists(p);
+}
+
+auto Util::getConfigFile(const fs::path& relativeFileName) -> fs::path {
+    fs::path p = getConfigSubfolder(relativeFileName.parent_path());
+    p /= relativeFileName.filename();
+    return p;
+}
+
+auto Util::getCacheFile(const fs::path& relativeFileName) -> fs::path {
+    fs::path p = getCacheSubfolder(relativeFileName.parent_path());
+    p /= relativeFileName.filename();
+    return p;
+}
+
+auto Util::getTmpDirSubfolder(const fs::path& subfolder) -> fs::path {
+    fs::path p(g_get_tmp_dir());
+    p /= FS(_F("xournalpp-{1}") % Util::getPid());
+    p /= subfolder;
+    return Util::ensureFolderExists(p);
+}
+
+auto Util::ensureFolderExists(const fs::path& p) -> fs::path {
+    if (g_mkdir_with_parents(p.u8string().c_str(), 0700) == -1) {
+        Util::execInUiThread([=]() {
+            string msg = FS(_F("Could not create folder: {1}") % p.string());
+            g_warning("%s", msg.c_str());
+            XojMsgBox::showErrorToUser(nullptr, msg);
+        });
+    }
+    return p;
```

Comment:
do this with filesystem functions.

Reply:
Are the default permissions used by fs::create_directories sufficient or should I force the 0700? fs::permissions can only manipulate one folder at a time, unfortunately.

Reply:
I think the default should be the same as the parent directory, so this should be fine.